[Docs] ppl documentation (#3715)
The-OpenROAD-Project/OpenROAD: src/ppl/README.md

## Commands

```{note}
- Parameters in square brackets `[-param param]` are optional.
- Parameters without square brackets `-param2 param2` are required.
```

### Define Pin Shape Pattern

The `define_pin_shape_pattern` command defines a pin placement grid on the
specified layer. This grid has positions inside the die area, not only at
the edges of the die boundary.

```tcl
define_pin_shape_pattern 
    [-layer layer]
    [-x_step x_step]
    [-y_step y_step]
    [-region {llx lly urx ury}]
    [-size {width height}]
    [-pin_keepout dist]
```

#### Options

| Switch Name | Description |
| ----- | ----- |

> | `-layer` | Defines a single top-most routing layer of the placement grid. |
> | `-x_step`, `-y_step` | Define the distance (in microns) between each valid position on the grid, in the x- and y-directions, respectively. |
> | `-region` | Defines the `{llx, lly, urx, ury}` region of the placement grid (in microns). |
> | `-size` | Defines the width and height (in microns) of the pins assigned to this grid. The centers of the pins are placed on the grid positions. Pins may have half of their shapes outside the defined region. |
> | `-pin_keepout` | Defines the boundary (in microns) around existing routing obstructions that the pins should avoid; this defaults to the `layer` minimum spacing. |

Reviewer: ```suggestion
| `-layer` | Defines a single top-most routing layer of the placement grid. |
| `-x_step`, `-y_step` | Define the distance (in microns) between each valid position on the grid in the x- and y-directions, respectively. |
| `-region` | Defines the `{llx, lly, urx, ury}` region of the placement grid (in microns). |
| `-size` | Defines the width and height (in microns) of the pins assigned to this grid. The centers of the pins are placed on the grid positions. Pins may have half of their shapes outside the defined region. |
| `-pin_keepout` | Defines the boundary (in microns) around existing routing obstructions that the pins should avoid; this defaults to the `layer` minimum spacing. |
```

layer with the minimum allowed spacing to facilitate 3DIC integration of
chips using face-to-face packaging technologies. These technologies include
[micro bumps](https://semiengineering.com/bumps-vs-hybrid-bonding-for-advanced-packaging/)
and
[hybrid bonding](https://www.3dincites.com/2018/04/hybrid-bonding-from-concept-to-commercialization/)
for high density face-to-face interconnect.

### Set IO Pin Constraint

The `set_io_pin_constraint` command sets region constraints for pins according
to the pin direction or the pin name. This command can be called multiple
times with different constraints.

You can use the `set_io_pin_constraint` command to restrict pins to the
pin placement grid created with the `define_pin_shape_pattern` command.

It is possible to use the `-region`, `-group` and `-order` arguments together
per `set_io_pin_constraint` call, but the `-mirrored_pins` argument should be
called alone.

```tcl
set_io_pin_constraint 
    [-direction direction]
    [-pin_names names]
    [-region edge:interval]
    [-mirrored_pins names]
    [-group]
    [-order]
```

#### Options

| Switch Name | Description |
| ----- | ----- |

> | `-direction` | Pin direction (input, output, inout, or feedthrough). |

Reviewer: ```suggestion
| `-direction` | Pin direction (`input`, `output`, `inout`, or `feedthrough`). |
```

Reviewer: I think you accendentally removed from the table.

> | `-pin_names` | List of names. Only one of these two (`-direction`, `-pin_names`) should be used in a single call for the `set_io_pin_constraint` command. |
> | `-region` | Syntax is `-region edge:interval`. The `edge` values are (top|bottom|left|right). The `interval` can be the whole edge, with the `*` value, or a range of values. |
> | `-mirrored_pins` | List of pins that sets pairs of pins that will be symmetrically placed in the vertical or the horizontal edges. The number of pins in this list **must be even**. For example, in `set_io_pin_constraint -mirrored_pins {pin1 pin2 pin3 pin4 pin5 pin6}`, the pins `pin1` and `pin2` will be placed symmetrically to each other. Same for `pin3` and `pin4`, and for `pin5` and `pin6`. |
> | `-group` | Flag places together on the die boundary the pin list defined in `-pin_names`, similar to the `-group_pins` option on `place_pins` command. |
> | `-order` | Flag places the pins ordered in ascending x/y position, and must be used only when `-group` is also used. |

Reviewer: ```suggestion
| `-pin_names` | List of names. Only one of (`-direction`, `-pin_names`) should be used in a single call for the `set_io_pin_constraint` command. |
| `-region` | Syntax is `-region edge:interval`. The `edge` values are (`top|bottom|left|right`). The `interval` can be the whole edge with the wildcard `*` value or a range of values. |
| `-mirrored_pins` | List of pins that sets pairs of pins that will be symmetrically placed in the vertical or the horizontal edges. The number of pins in this list **must be even**. For example, in `set_io_pin_constraint -mirrored_pins {pin1 pin2 pin3 pin4 pin5 pin6}`, the pins `pin1` and `pin2` will be placed symmetrically to each other. Same for `pin3` and `pin4`, and for `pin5` and `pin6`. |
| `-group` | Flag places together on the die boundary the pin list defined in `-pin_names,` similar to the `-group_pins` option on the `place_pins` command. |
| `-order` | Flag places the pins ordered in ascending x/y position and must be used only when `-group` is also used. |
```

| `-corner_avoidance` | The distance (in microns) from each corner within which pin placement should be avoided. |
| `-min_distance` | The minimum distance between pins on the die boundary. This distance can be in microns (default) or in number of tracks between each pin. |
| `-min_distance_in_tracks` | Flag that allows setting the min distance in number of tracks instead of microns. |
| `-exclude` | A region where pins cannot be placed. Either `top|bottom|left|right:edge_interval`, which is the edge interval from the selected edge; `begin:end` for begin-end of all edges. |
| `-group_pins` | A list of pins to be placed together on the die boundary. |
| `-annealing` | Flag to enable simulated annealing pin placement. |

The `exclude` option syntax is `-exclude edge:interval`. The `edge` values are
(top|bottom|left|right). The `interval` can be the whole edge, with the `*`
value, or a range of values. For example, in `place_pins -hor_layers metal2
-ver_layers metal3 -exclude top:* -exclude right:15-60.5 -exclude left:*-50`
three intervals are excluded: the whole top edge, the right edge from 15
microns to 60.5 microns, and the left edge from its beginning to 50 microns.

> ### Useful developer functions

Reviewer: ```suggestion
### Useful Developer Commands
```

If you are a developer, you might find these useful. More details can be found in the [source file](./src/IOPlacer.cpp) or the [swig file](./src/IOPlacer.i).

> | Function Name | Description |

Reviewer: ```suggestion
| Command Name | Description |
```

| ----- | ----- |
| `parse_edge` | Parse edge (top/bottom/left/right). |
| `parse_direction` | Parse direction. |
| `parse_excludes_arg` | Parse excluded arguments. |
| `parse_group_pins_arg` | Parse group pins arguments. | 
| `parse_layer_name` | Parse layer name. |
| `parse_pin_names` | Parse pin names. |
| `get_edge_extreme` | Get extremes of edge. |
| `exclude_intervals` | Set exclude interval. |
| `add_pins_to_constraint` | Add pins to constrained region. |
| `add_pins_to_top_layer` | Add pins to top layer. | 

### Useful developer arguments

| Switch Name | Description |
| ----- | ----- |
| `-random_seed` | Specify the seed for random operations. |

> | `-random` | When this flag is enabled, the pin placement is random. |

Reviewer: The same issue as with the other tool, it is not clear which command these flags can be used with.

The `up` option is only available when the pin placement grid is created with
the `define_pin_shape_pattern` command.

### Clear IO Pin Constraints

The `clear_io_pin_constraints` command clears all the previously-defined
constraints and pin shape patterns created with `set_io_pin_constraint` or
`define_pin_shape_pattern`.

```tcl
clear_io_pin_constraints
```

### Set Pin Length

The `set_pin_length` command defines the length of all vertical and horizontal
pins.

```tcl
set_pin_length 
    [-hor_length h_length]
    [-ver_length v_length]
```

#### Options

| Switch Name | Description |
| ----- | ----- |

> | `-hor_length` | Defines the length (in microns) of the horizontal pins. |
> | `-ver_length` | Defines the length (in microns) of the vertical pins. |
> 

Reviewer: ```suggestion
| `-hor_length` | The length (in microns) of the horizontal pins. |
| `-ver_length` | The length (in microns) of the vertical pins. |

```

### Set Pin Extension

The `set_pin_length_extension` command defines the an extension of the length
of all vertical and horizontal pins. Note that this command may generate pins
partially outside the die area.

```tcl
set_pin_length_extension 
    [-hor_extension h_extension]
    [-ver_extension v_extension]
```

#### Options

| Switch Name | Description |
| ----- | ----- |

> | `-hor_extension` | Defines the extension length (in microns) for the horizontal pins. |
> | `-ver_extension` | Defines the extension length (in microns) for the vertical pins. |

Reviewer: ```suggestion
| `-hor_extension` | The length (in microns) for the horizontal pins extension. |
| `-ver_extension` | The length (in microns) for the vertical pins extension. |
```

### Set Pin Thick Multiplier

The `set_pin_thick_multiplier` command defines a multiplier for the thickness of all
vertical and horizontal pins.

```tcl
set_pin_thick_multiplier 
    [-hor_multiplier h_mult]
    [-ver_multiplier v_mult]
```

#### Options

| Switch Name | Description |
| ----- | ----- |

> | `-hor_multiplier` | Defines the thickness multiplier for the horizontal pins. |
> | `-ver_multiplier` | Defines the thickness multiplier for the vertical pins. |

Reviewer: ```suggestion
| `-hor_multiplier` | The thickness multiplier for the horizontal pins. |
| `-ver_multiplier` | The thickness multiplier for the vertical pins. |
```

### Set Simulated Annealing Parameters

The `set_simulated_annealing` command defines the parameters for simulated annealing pin placement.

```tcl
set_simulated_annealing
    [-temperature temperature]
    [-max_iterations iter]
    [-perturb_per_iter perturbs]
    [-alpha alpha]
```

#### Options

| Switch Name | Description |
| ----- | ----- |

> | `-temperature` | Specify the temperature parameter, must be positive. |
> | `-max_iterations` | Specify the maximum number of iterations. |
> | `-perturb_per_iter` | Specify the number of perturbations per iteration. |
> | `-alpha` | Specify the temperature decay factor, must be positive. |

Reviewer: ```suggestion
| `-temperature` | Temperature parameter, must be positive. |
| `-max_iterations` | The maximum number of iterations. |
| `-perturb_per_iter` | The number of perturbations per iteration. |
| `-alpha` | The temperature decay factor, must be positive. |
```

Reviewer: Check data type (int, float, ...) and add relevant default/allowed range text.

### Place Individual Pin

> The `place_pin` command places a specific pin in the specified location, with the specified size.
> It is recommended that the placement of individual pins be done before the `place_pins` command,
> as the routing tracks occupied by these individual pins will be blocked, preventing overlaps.

Reviewer: ```suggestion
The `place_pin` command places a specific pin in the specified location with the specified size.
It is recommended that individual pins be placed before the `place_pins` command,
as the routing tracks occupied by these individual pins will be blocked, preventing overlaps.
```

To place an individual pin:

```tcl
place_pin 
    [-pin_name pin_name]
    [-layer layer]
    [-location {x y}]
    [-pin_size {width height}]
    [-force_to_die_boundary]
```

#### Options

| Switch Name | Description |
| ----- | ----- |

> | `-pin_name` | Option is the name of a pin of the design. |
> | `-layer` | Defines the routing layer where the pin is placed. |
> | `-location` | Defines the center of the pin (in microns). |
> | `-pin_size` | Option defines the width and height of the pin (in microns). |
> | `-force_to_die_boundary` | When this flag is enabled, the pin will be snapped to the nearest routing track, next to the die boundary. |

Reviewer: ```suggestion
| `-pin_name` | The name of a pin of the design. |
| `-layer` | The routing layer where the pin is placed. |
| `-location` | The center of the pin (in microns). |
| `-pin_size` | Tthe width and height of the pin (in microns). |
| `-force_to_die_boundary` | When this flag is enabled, the pin will be snapped to the nearest routing track, next to the die boundary. |
```

### Place All Pins

Use the following command to perform pin placement:

```tcl
place_pins 
    [-hor_layers h_layers]
    [-ver_layers v_layers]
    [-random_seed seed]
    [-random]
    [-corner_avoidance length]
    [-min_distance distance]
    [-min_distance_in_tracks]
    [-exclude region]
    [-group_pins pin_list]
    [-annealing]
```

#### Options

| Switch Name | Description |
| ----- | ----- |

> | `-hor_layers` | Specify the layers to create the metal shapes of pins placed in horizontal tracks. Can be a single layer or a list of layer names. |
> | `-ver_layers` | Specify the layers to create the metal shapes of pins placed in vertical tracks. Can be a single layer or a list of layer names. |

Reviewer: ```suggestion
| `-hor_layers` | Specify the layers to create the metal shapes of pins placed in horizontal tracks. It can be a single layer or a list of layer names. |
| `-ver_layers` | Specify the layers to create the metal shapes of pins placed in vertical tracks. It can be a single layer or a list of layer names. |
```

| `-random_seed` | Specify the seed for random operations. |

> | `-random` | When this flag is enabled, the pin placement is random. |

Reviewer: Developer?

| `-corner_avoidance` | Specify the distance (in microns) from each corner within which pin placement should be avoided. |

> | `-min_distance` | Specify the minimum distance between pins on the die boundary. This distance can be in microns (default) or in number of tracks between each pin. |

Reviewer: Remove "Specify" and "Defines the" statements.

value, or a range of values. For example, in `place_pins -hor_layers metal2
-ver_layers metal3 -exclude top:* -exclude right:15-60.5 -exclude left:*-50`
three intervals are excluded: the whole top edge, the right edge from 15
microns to 60.5 microns, and the left edge from its beginning to 50 microns.

### Useful developer functions

If you are a developer, you might find these useful. More details can be found in the [source file](./src/IOPlacer.cpp) or the [swig file](./src/IOPlacer.i).

```tcl
# parse edge/direction/exclude args/group pins arg/layer name/pin name
parse_edge cmd edge
parse_direction cmd direction
parse_excludes_arg args_var
parse_group_pins_arg args_var
parse_layer_name layer_name
parse_pin_names cmd names

# get extremes of edge
get_edge_extreme cmd begin edge

# set exclude interval
exclude_intervals cmd intervals

# add pins to constrained region
add_pins_to_constraint cmd names edge begin end edge_name

# add pins to top layer
add_pins_to_top_layer cmd names llx lly urx ury
```

## Example scripts

Example scripts of IOPlacer running on a sample design of `gcd` as follows:

```tcl
./test/gcd.tcl
```

## Regression tests

> There are a set of regression tests in `/test`. For more information, refer to this [section](../../README.md#regression-tests).

Reviewer: ```suggestion
There are a set of regression tests in `./test`. For more information, refer to this [section](../../README.md#regression-tests).
```

To place an individual pin:

```tcl
place_pin 
    [-pin_name pin_name]
    [-layer layer]
    [-location {x y}]
    [-pin_size {width height}]
    [-force_to_die_boundary]
```

#### Options

| Switch Name | Description |
| ----- | ----- |
| `-pin_name` | The name of a pin of the design. |
| `-layer` | The routing layer where the pin is placed. |
| `-location` | The center of the pin (in microns). |
| `-pin_size` | Tthe width and height of the pin (in microns). |
| `-force_to_die_boundary` | When this flag is enabled, the pin will be snapped to the nearest routing track, next to the die boundary. |

### Place All Pins

Use the following command to perform pin placement:

```tcl
place_pins 
    [-hor_layers h_layers]
    [-ver_layers v_layers]
    [-random_seed seed]
    [-random]
    [-corner_avoidance length]
    [-min_distance distance]
    [-min_distance_in_tracks]
    [-exclude region]
    [-group_pins pin_list]
    [-annealing]
```

#### Options

| Switch Name | Description |
| ----- | ----- |
| `-hor_layers` | The layers to create the metal shapes of pins placed in horizontal tracks. It can be a single layer or a list of layer names. |

> | `-ver_layers` | The layers to create the metal shapes of pins placed in vertical tracks. It can be a single layer or a list of layer names. |
> | `-random_seed` | Refer to the developer arguments [here](#developer-arguments). |
> | `-random` | Refer to the developer arguments [here](#developer-arguments). |
> | `-corner_avoidance` | The distance (in microns) from each corner within which pin placement should be avoided. |

Reviewer: ```suggestion
| `-ver_layers` | The layers to create the metal shapes of pins placed in vertical tracks. It can be a single layer or a list of layer names. |
| `-corner_avoidance` | The distance (in microns) from each corner within which pin placement should be avoided. |
```

| `-min_distance` | The minimum distance between pins on the die boundary. This distance can be in microns (default) or in number of tracks between each pin. |
| `-min_distance_in_tracks` | Flag that allows setting the min distance in number of tracks instead of microns. |
| `-exclude` | A region where pins cannot be placed. Either `top|bottom|left|right:edge_interval`, which is the edge interval from the selected edge; `begin:end` for begin-end of all edges. |
| `-group_pins` | A list of pins to be placed together on the die boundary. |
| `-annealing` | Flag to enable simulated annealing pin placement. |

The `exclude` option syntax is `-exclude edge:interval`. The `edge` values are
(top|bottom|left|right). The `interval` can be the whole edge, with the `*`
value, or a range of values. For example, in `place_pins -hor_layers metal2
-ver_layers metal3 -exclude top:* -exclude right:15-60.5 -exclude left:*-50`
three intervals are excluded: the whole top edge, the right edge from 15
microns to 60.5 microns, and the left edge from its beginning to 50 microns.

> ### Developer Arguments

Reviewer: ```suggestion
#### Developer Arguments
```
